grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -136,6 +175,122 @@ public static GraqlTraversal createTraversal(PatternAdmin pattern, EmbeddedGrakn
         return plan;
     }
 
+    private static void inferRelationshipTypes(EmbeddedGraknTx<?> tx, Set<Fragment> allFragments) {
+        // find all vars representing types
+        Map<Var, Type> labelVarTypeMap = new HashMap<>();
+        allFragments.stream()
+                .filter(LabelFragment.class::isInstance)
+                .forEach(fragment -> {
+                    // TODO: labels() should return ONE label instead of a set
+                    Type type = tx.getType(Iterators.getOnlyElement(((LabelFragment) fragment).labels().iterator()));
+                    if (type != null && !type.isRole()) labelVarTypeMap.put(fragment.start(), type);
```

Comment:
It's actually possible!

Reply:
Is it OK if it just overrides whatever is there?

Reply:
It's kind of okay, but it would be nice to work out which label is the subtype:
`$x label person, $x label mammal;`
Or figure our it wouldn't return anything:
`$x label person, $x label dog;`